c7n/resources/s3.py

```diff
@@ -920,6 +936,85 @@ def process(self, buckets):
         return None
 
 
+@actions.register('set-statements')
+class SetPolicyStatement(BucketActionBase):
+    """Action to add or update policy statements to S3 buckets
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: force-s3-https
+                resource: s3
+                actions:
+                  - type: set-statements
+                    statements:
+                      - Sid: "DenyHttp"
+                        Effect: "Deny"
+                        Action: "s3:GetObject"
+                        Principal:
+                          AWS: "*"
+                        Resource: "arn:aws:s3:::{bucket_name}/*"
+                        Condition:
+                          Bool:
+                            "aws:SecureTransport": false
+    """
+
+    permissions = ('s3:PutBucketPolicy',)
+
+    schema = type_schema(
+        'set-statements',
+        **{
+            'statements': {
+                'type': 'array',
+                'required': True,
+                'items': {
+                    'type': 'object',
+                    'properties': {
+                        'Sid': {'type': 'string'},
+                        'Effect': {'type': 'string', 'enum': ['Allow', 'Deny']},
+                        'Principal': {'anyOf': [{'type': 'object'}, {'type': 'array'}]},
+                        'NotPrincipal': {'anyOf': [{'type': 'object'}, {'type': 'array'}]},
+                        'Action': {'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                        'NotAction': {'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                        'Resource': {'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                        'NotResource': {'anyOf': [{'type': 'string'}, {'type': 'array'}]},
+                        'Condition': {'type': 'object'}
+                    },
+                    'required': ['Sid', 'Effect'],
+                    'oneOf': [
+                        {'required': ['Action', 'Resource']},
+                        {'required': ['NotAction', 'Resource']},
+                        {'required': ['Action', 'NotResource']},
+                        {'required': ['NotAction', 'NotResource']}
+                    ]
+                }
+            }
+        }
+    )
+
+    def process_bucket(self, bucket):
+        policy = bucket.get('Policy')
+        if policy is None:
```

Comment:
if policy is none refusing to set a new policy seems odd, you can check denied annotation if thats the consideration.